File: jps/src/jpsutil/jpsutil.go
/*

	JOTPOT Server
	Version 26B-0

	Copyright (c) 2016-2017 Jacob O'Toole

	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.

*/

package jpsutil

import (
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"runtime"
)

//GetNodePath checks PATH env varaible and cwd/node for the node binary and returns the path of it.
func GetNodePath() string {
	var thisPath string
	var err error
	commands := []string{"nodejs", "node"}
	for _, command := range commands {
		if runtime.GOOS == "windows" {
			command += ".exe"
		}
		for _, p := range filepath.SplitList(os.Getenv("PATH")) {
			thisPath = filepath.Join(p, command)
			_, err = os.Stat(thisPath)
			if err != nil {
				continue
			}
			return thisPath
		}
		thisPath, err = os.Getwd()
		thisPath = filepath.Join(thisPath, "node", command)
		_, err = os.Stat(thisPath)
		if err != nil {
			continue
		}
		return thisPath
	}
	fmt.Fprintln(os.Stderr, "No Node.js binary")
	os.Exit(1)
	panic("No Node.js binary")
}

//Uint32ToBytes creates a slice of 4 bytes as a representation of a uint32
func Uint32ToBytes(in uint32) []byte {
	return []byte{byte(in >> 24), byte((in >> 16) & 255), byte((in >> 8) & 255), byte(in & 255)}
}

//Uint32FromBytes creates a uint32 from a slice of 4 bytes
func Uint32FromBytes(in []byte) uint32 {
	return (uint32(in[0]) << 24) + (uint32(in[1]) << 16) + (uint32(in[2]) << 8) + uint32(in[3])
}

//GetData takes a connection and reads the specified amout of bytes and returns them in a slice
func GetData(con net.Conn, toGet int) (out []byte) {
	var got int
	var buff []byte
	if toGet < 1024 {
		buff = make([]byte, toGet)
	} else {
		buff = make([]byte, 1024)
	}
	var n int
	var err error
	for got < toGet {
		if toGet-got < len(buff) {
			buff = buff[:toGet-got]
		}
		n, err = con.Read(buff)
		if err != nil && err != io.EOF {
			panic(err)
		} else if n == 0 && err == io.EOF {
			panic(err)
		}
		got += n
		out = append(out, buff[:n]...)
	}
	return
}

//GetDataPre is the same as GetData but uses predeclared variables
func GetDataPre(con net.Conn, toGet int, got, n *int, err *error, buff, out *[]byte) {
	*got = 0
	*out = []byte{}
	if toGet < 1024 {
		*buff = make([]byte, toGet)
	} else {
		*buff = make([]byte, 1024)
	}
	*n = 0
	for *got < toGet {
		if toGet-*got < len(*buff) {
			*buff = (*buff)[:toGet-*got]
		}
		*n, *err = con.Read(*buff)
		if *err != nil && *err != io.EOF {
			panic(*err)
		} else if *n == 0 && *err == io.EOF {
			panic(err)
		}
		*got += *n
		*out = append(*out, (*buff)[:*n]...)
	}
}

//GetByte reads 1 byte from the con
func GetByte(con net.Conn) (out byte) {
	var n int
	var err error
	buff := []byte{0}
	for n < 1 {
		n, err = con.Read(buff)
		if err != nil {
			panic(err)
		}
	}
	return buff[0]
}

//GetBytePre reads 1 byte from the con
//but uses already declared variables
func GetBytePre(con net.Conn, n *int, err *error, buff *[]byte) {
	*n = 0
	*buff = []byte{0}
	for *n < 1 {
		*n, *err = con.Read(*buff)
		if *err != nil {
			panic(*err)
		}
	}
}

//CheckAddr checks if an address exists, if so, it returns true. If not, false.
func CheckAddr(addr string) bool {
	con, err := net.Dial("tcp", addr)
	if err != nil {
		return false
	}
	con.Close()
	return true
}

//CopyFile copies file p1 to p2
func CopyFile(p1, p2 string) error {
	src, err := os.Open(p1)
	if err != nil {
		return err
	}
	dest, err := os.OpenFile(p2, os.O_WRONLY|os.O_CREATE, 0664)
	if err != nil {
		return err
	}
	stats, err := src.Stat()
	if err != nil {
		return err
	}
	err = dest.Truncate(stats.Size())
	if err != nil {
		return err
	}
	buff := make([]byte, 1024)
	var n int
	var err2 error
	for {
		n, err = src.Read(buff)
		if err != nil && err != io.EOF {
			return err
		}
		_, err2 = dest.Write(buff[:n])
		if err2 != nil {
			return err2
		}
		if err == io.EOF {
			break
		}
	}
	if err = src.Close(); err != nil {
		return err
	}
	if err = dest.Close(); err != nil {
		return err
	}
	return nil
}

//Search searches the give slice for the string and returns the index, or -1 if it isn't found
func Search(slice []string, want string) int {
	for i, val := range slice {
		if val == want {
			return i
		}
	}
	return -1
}

//DaemonOnlyArgs are the arguments that if handles by the daemon should not be passed on to the server
var DaemonOnlyArgs = []string{"-autoreload", "-stayalive", "-keepalive"}

//Args type is a represntation of the server arguments
type Args []string

//ToServer returns the args that should be passed on to the server
func (args Args) ToServer() (out Args) {
	var isToServer bool
	for _, arg := range args {
		if arg[0] == 45 {
			isToServer = Search(DaemonOnlyArgs, arg) == -1
		}
		if isToServer {
			out = append(out, arg)
		}
	}
	return
}

//Has returns a boolean representing of the given argument is given
func (args Args) Has(arg string) bool {
	for _, ta := range args {
		if ta == arg {
			return true
		}
	}
	return false
}

File: jps-util.js
/*
	
	JOTPOT Server
	Version 27A-0
	
	Copyright (c) 2016-2017 Jacob O'Toole
	
	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:
	
	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.
	
	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.
	
*/

"use strict";

//Modules
const fs = require("fs") ;
const path = require("path") ;
const os = require("os") ;
const http = require("http") ;

//Default configuration
const defaultConfig = {
	
	"dataPort": 500,
	"controlers":["::1","127.0.0.1","::ffff:127.0.0.1"],
	
	"httpServers": [
		{
			"port": 80,
			"websocket": false
		}
	],
	"httpsServers": [],
	
	"redirectToHttps": [],
	"mustRedirectToHttps": [],
	"dontRedirect": [],
	
	"hostRedirects":{},
	"hostnameRedirects":{},
	"hostAlias":{},
	"hostnameAlias":{},
	"pageAlias":{},
	
	"addVarsByDefault": false,
	"doVarsForIfNotByDefault": [],
	
	"cache": [],
	
	"errorTemplate": "errorTemp.jpt",
	
	"defaultHost": "default:0",
	"useDefaultHostIfHostDoesNotExist": true,
	
	"behindLoadBalencer": false,
	"fallbackToNoPort": true,
	
	"defaultHeaders": {},
	
	"CORS":[],
	
	"enableLearning": true,
	
	"threads": 0
	
} ;

const loadConfigFile = p => {
	if (fs.existsSync(p)) {
		return fs.readFileSync(p).toString() ;
	}
	return fs.readFileSync(path.join(__dirname, p)).toString() ;
} ;
const doesConfigFileExist = p => (fs.existsSync(p) || fs.existsSync(path.join(__dirname, p))) ;

//Load the comfig and fill in any blanks. If it doesn't exist, set the config to the default config.
function loadConfig() {
	let config ;
	//If it exists, load it, parse it and fill in any blanks or throw if the types aren't correct
	if (doesConfigFileExist("config.json")) {
		config = loadConfigFile("config.json") ;
		try {
			config = JSON.parse(config) ;
		} catch (err) {
			console.warn("Error parsing config.json!") ;
			console.info("Error parsing config.json!") ;
			console.warn(err) ;
			console.info(err) ;
			console.warn("Exiting") ;
			console.info("Exiting") ;
			process.exit(1) ;
		}
		for (let doing in defaultConfig) {
			if (typeof config[doing] === "undefined") {
				config[doing] = defaultConfig[doing] ;
			} else if (typeof config[doing] !== typeof defaultConfig[doing]) {
				throw new Error(`The ${doing} property in config.json must be of type ${typeof defaultConfig[doing]}.`) ;
			}
		}
	} else {
		console.warn("Config file does not exist, using default config.") ;
		config = new Object() ;
		Object.assign(config, defaultConfig) ;
	}
	return config ;
}

//Returns a promise that resolves with the contents of the given requests body.
function getData(req) {
	return new Promise((resolve, reject)=>{
		//If the request body has a specified length
		if (typeof req.headers["content-length"] !== "undefined") {
			//Parse and reject if not number
			let dLength = parseInt(req.headers["content-length"], 10) ;
			if (isNaN(dLength)) {
				reject(new Error("Content-Length header must be a number")) ;
				return ;
			}
			//Create the buffer
			let data = Buffer.alloc(dLength) ;
			let currentPos = 0 ;
			let errorSent = false ;
			req.on("data", d=>{
				//If we have more data than we can take, reject.
				if (currentPos + d.length > data.length) {
					//But only reject once
					if (errorSent) {
						return ;
					}
					errorSent = true ;
					reject(new Error("Request body was longer than the Content-Length header.")) ;
					return ;
				}
				//Write the recieved data to the output and increment the curret position by the amount of data copied.
				currentPos += d.copy(data, currentPos) ;
			}) ;
			//Resolve when the request has ended.
			req.on("end", ()=>{
				resolve(data) ;
			}) ;
			return ;
		}
		//No idea how much data, so we have to concatinate all the Buffers :(
		let data = Buffer.alloc(0) ;
		req.on("data", d=>{
			data = Buffer.concat([data, d], data.length + d.length) ;
		}) ;
		//Resolve when we are done
		req.on("end", ()=>{
			resolve(data) ;
		}) ;
	}) ;
}

function coughtError(err, where="", resp, rID="", userMessage="") {
	let isUnknown = false ;
	console.warn("---------------") ;
	if (err && err.stack) {
		console.warn("!!! Error" + where + ":") ;
		console.warn("\t" + err.stack.replace(/\n/g,"\n\t")) ;
	} else if (err) {
		console.warn("!!! Error" + where + ":") ;
		console.warn("\t" + err.replace(/\n/g,"\n\t")) ;
	} else {
		console.warn("!!! Error" + where + ", details unknown. Stack unavailable.") ;
		isUnknown = true ;
	}
	console.warn("---------------") ;
	if (resp) {
		try {
			if (userMessage) {
				module.exports.sendError(500, userMessage, resp, rID) ;
			} else {
				module.exports.sendError(500, `A${isUnknown?"n unknown":" known "} error occured.${isUnknown?"":" I just don't want to tell you what went wrong. Nothing personal, honestly! It's not like I don't trust you."}.`, resp, rID) ;
			}
		} catch (err) {
			coughtError(err, " sending error response") ;
		}
	}
}

function getFilePath(req) {
	return path.normalize((req.url.usePortInDirectory?req.url.host:req.url.hostname).replace(":", ";") + req.url.pathname) ;
}

function escapeHTML(text) {
	return text
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&#34;")
		.replace(/'/g, "&#39;") ;
}

function getTime() {
	let time = process.hrtime() ;
	return time[0] * 10e3 + time[1] * 10e-6 ;
}

class time {
	constructor(total, fileOp) {
		this.total = total ;
		this.fileOp = fileOp ;
		this.process = total - fileOp ;
	}
}

class timer {
	constructor() {
		this.fileOp = false ;
		this.fileOpStart = 0 ;
		this.fileOpTime = 0 ;
		this.start = getTime() ;
	}
	startFileOp() {
		if (!this.fileOp) {
			this.fileOp = true ;
			this.fileOpStart = getTime() ;
		}
	}
	stopFileOp() {
		if (this.fileOp) {
			this.fileOpTime += getTime() - this.fileOpStart ;
			this.fileOp = false ;
		}
	}
	currentTime() {
		const ct = getTime() ;
		if (this.fileOp) {
			this.fileOpTime += ct - this.fileOpStart ;
			this.fileOpStart = ct ;
		}
		return new time(ct - this.start, this.fileOpTime) ;
	}
}

function getLogDate(d=new Date()) {
	return `${d.getUTCFullYear()}-${d.getUTCMonth()}-${d.getUTCDate()}` ;
}

function getLogTime(d=new Date()) {
	return `${d.getUTCHours()}-${d.getUTCMinutes()}-${d.getUTCSeconds()}.${d.getUTCMilliseconds()/1000}` ;
}

function startLog() {
	const d = new Date() ;
	return `#Version: 1.0${os.EOL}#Software: JOTPOT Server${os.EOL}#Start-Date: ${getLogDate(d)} ${getLogTime(d)}${os.EOL}#Fields: ${fields}` ;
}

function updateLog() {
	const d = new Date() ;
	return `#Date: ${getLogDate(d)} ${getLogTime(d)}` ;
}

function log(req, status, location, cached, learned) {
	console.serverlog(writeLog(req, status, location, cached, learned)) ;
}

const fields = [
	"r-id",
	"c-ip",
	"c-ip-forwarded",
	"c-id",
	"c-username",
	"recieved-date",
	"recieved-time",
	"date",
	"time",
	"time-taken",
	"time-taken-process",
	"time-taken-disk",
	"cs-method",
	"cs-uri",
	"cs-uri-stem",
	"cs-uri-query",
	"cs-url",
	"sc-url-served",
	"cs-requested",
	"sc-served",
	"cs-http-version",
	"cs(User-Agent)",
	"cs(Referer)",
	"cached",
	"learned",
	"sc-status",
	"sc-comment",
	"sc-location"
].join(" ") ;

function writeLog(req, status, location, cached, learned) {
	const d = new Date() ;
	const t = req.timer.currentTime() ;
	return [
		req.jpid, //r-id
		req.remoteAddress, //c-ip
		req.ip, //c-ip-forward
		req.user||'-', //c-id
		'-', //c-username
		getLogDate(req.time), //recieved-date
		getLogTime(req.time), //recieved-time
		getLogDate(d), //date
		getLogTime(d), //time
		t.total.toString(10), //time-taken
		t.process.toString(10), //time-taken-process
		t.fileOp.toString(10), //time-taken-disk
		req.method, //cs-method
		req.orig_url.path, //cs-uri
		req.orig_url.pathname, //cs-uri-stem
		req.orig_url.query, //cs-uri-query
		req.orig_url.href, //cs-url
		req.url.href, //sc-url-served
		req.orig_url.fullvalue, //cs-requested
		req.url.fullvalue, //cs-served
		req.httpVersion, //cs-http-version
		req.headers["user-agent"]||'-', //cs(User-Agent)
		req.headers["referer"]||'-', //cs(Referer)
		cached, //cached
		learned, //leaned
		status.toString(10), //sc-status
		http.STATUS_CODES[status], //sc-comment
		location //sc-location
	].join(' ') ;
}

module.exports = {
	coughtError,
	doesConfigFileExist,
	escapeHTML,
	getData,
	getFilePath,
	loadConfig,
	loadConfigFile,
	log,
	startLog,
	updateLog,
	sendError: ()=>{},
	timer
} ;
